File: domains/finance/projects/stock-market/development/software/ios/StockPrediction/Prediction/PredictionExplainer.swift
import Foundation

/// One feature's marginal contribution to the calibrated up-probability,
/// measured by replacing that feature with its neutral training baseline.
struct FeatureContribution {
    let feature: String
    let label: String
    let value: Float
    /// score(features) - score(features with this feature set to baseline).
    /// Positive => this feature pushes the prediction UP relative to neutral.
    let delta: Double
    var pushesUp: Bool { delta >= 0 }
}

/// Perturbation-based "why this prediction" explainer.
///
/// For each of the 16 features it asks: how would the calibrated up-probability
/// change if this single feature were at its neutral baseline instead of its
/// actual value? The signed change is the feature's contribution. This is a
/// simple, model-agnostic local attribution that reuses the live ONNX session
/// through the supplied `score` closure (which must return a CALIBRATED probUp).
enum PredictionExplainer {

    /// Computes per-feature contributions, sorted by |delta| descending.
    /// - Parameters:
    ///   - features: the 16-feature vector (MOBILE_FEATURES order).
    ///   - horizon: unused for scoring here (the closure already bakes it in),
    ///     kept for call-site clarity and future use.
    ///   - score: closure returning the calibrated probUp for a feature vector.
    static func contributions(
        features: [Float],
        horizon: String,
        score: ([Float]) -> Double
    ) -> [FeatureContribution] {
        let names = ModelMeta.shared.features
        guard !names.isEmpty, features.count == names.count else { return [] }

        let base = score(features)
        var result: [FeatureContribution] = []
        result.reserveCapacity(names.count)

        for i in 0..<names.count {
            let name = names[i]
            var perturbed = features
            perturbed[i] = ModelMeta.shared.baseline(for: name)
            let neutral = score(perturbed)
            let delta = base - neutral
            result.append(
                FeatureContribution(
                    feature: name,
                    label: ModelMeta.shared.label(for: name),
                    value: features[i],
                    delta: delta
                )
            )
        }

        return result.sorted { abs($0.delta) > abs($1.delta) }
    }

    /// One plain-English sentence describing the top 2-3 drivers and whether
    /// each supports or opposes the predicted direction.
    static func rationale(direction: String, contributions: [FeatureContribution]) -> String {
        let up = direction.uppercased() == "UP"
        let leaning = up ? "Leaning UP" : "Leaning DOWN"

        let top = Array(contributions.prefix(3)).filter { abs($0.delta) > 1e-6 }
        guard !top.isEmpty else {
            return "\(leaning), with no single feature standing out."
        }

        // A feature "supports" the predicted direction when its push matches it.
        func supports(_ c: FeatureContribution) -> Bool { c.pushesUp == up }

        let supporting = top.filter(supports).map { $0.label.lowercasedFirst() }
        let opposing = top.filter { !supports($0) }.map { $0.label.lowercasedFirst() }

        var sentence = leaning
        if !supporting.isEmpty {
            sentence += " mainly because \(joinClause(supporting)) "
            sentence += supporting.count == 1 ? "is favorable" : "are favorable"
        } else {
            // All top drivers oppose the lean (e.g. a weak/forced call).
            sentence += " despite \(joinClause(opposing)) "
            sentence += opposing.count == 1 ? "pulling the other way" : "pulling the other way"
            return sentence + "."
        }

        if !opposing.isEmpty {
            sentence += ", partly offset by \(joinClause(opposing))"
        }
        return sentence + "."
    }

    /// "a", "a and b", or "a, b and c".
    private static func joinClause(_ items: [String]) -> String {
        switch items.count {
        case 0: return ""
        case 1: return items[0]
        case 2: return "\(items[0]) and \(items[1])"
        default:
            let head = items.dropLast().joined(separator: ", ")
            return "\(head) and \(items.last!)"
        }
    }
}

private extension String {
    /// Lowercases only the first character (keeps acronyms like "RSI" intact
    /// beyond the first letter), for mid-sentence use.
    func lowercasedFirst() -> String {
        guard let first = first else { return self }
        return first.lowercased() + dropFirst()
    }
}

File: domains/finance/projects/stock-market/development/software/android/core/src/main/java/com/prediction/stockmarket/prediction/PredictionExplainer.kt
package com.prediction.stockmarket.prediction

import kotlin.math.abs

/**
 * Perturbation-based "why this prediction" explainer.
 *
 * For each input feature, it replaces that feature with its training baseline and
 * measures how the CALIBRATED P(up) changes. A large positive delta means the
 * feature's actual value pushes the probability up relative to a neutral input;
 * a large negative delta means it pushes it down.
 */
object PredictionExplainer {

    data class FeatureContribution(
        val feature: String,
        val label: String,
        val value: Float,
        val delta: Double,
    ) {
        val pushesUp get() = delta >= 0
    }

    /**
     * @param features the 16-feature input (MOBILE_FEATURES order, newest-first derived).
     * @param horizon  UI horizon string (used only for labelling; scoring already encodes it).
     * @param score    closure returning the CALIBRATED P(up) for a given feature array.
     * @return contributions sorted by absolute impact, descending.
     */
    fun explain(
        features: FloatArray,
        horizon: String,
        score: (FloatArray) -> Double,
    ): List<FeatureContribution> {
        val base = score(features)
        val names = ModelMeta.featureNames
        val out = ArrayList<FeatureContribution>(features.size)
        for (i in features.indices) {
            val name = if (i < names.size) names[i] else "feature_$i"
            val perturbed = features.copyOf()
            perturbed[i] = ModelMeta.baseline(name)
            val delta = base - score(perturbed)
            out.add(
                FeatureContribution(
                    feature = name,
                    label = ModelMeta.label(name),
                    value = features[i],
                    delta = delta,
                )
            )
        }
        return out.sortedByDescending { abs(it.delta) }
    }

    /**
     * One plain-English sentence summarising the top drivers and whether each
     * supports or opposes the predicted [direction] ("UP"/"DOWN").
     */
    fun rationale(direction: String, contributions: List<FeatureContribution>): String {
        if (contributions.isEmpty()) return "No individual driver stands out for this prediction."
        val up = direction.equals("UP", ignoreCase = true)
        val top = contributions.take(3).filter { abs(it.delta) > 1e-6 }
        if (top.isEmpty()) return "No individual driver stands out for this prediction."

        val parts = top.map { c ->
            // A contribution supports the direction if it pushes the same way as the prediction.
            val supports = c.pushesUp == up
            val verb = if (supports) "supports" else "weighs against"
            "${c.label} $verb it"
        }
        val joined = when (parts.size) {
            1 -> parts[0]
            2 -> "${parts[0]} and ${parts[1]}"
            else -> "${parts[0]}, ${parts[1]}, and ${parts[2]}"
        }
        val dirWord = if (up) "higher" else "lower"
        return "The model leans $dirWord mainly because $joined."
    }
}
